File: latest_ran_output/gpt_oss_120b/render_005_security_aware.js
const express = require('express');
const bodyParser = require('body-parser');
const bcrypt = require('bcrypt');
const jwt = require('jsonwebtoken');
const url = require('url');

const app = express();
app.use(bodyParser.json());

// Mock user store (username => bcrypt hash)
const USERS = {
  // password: secret123
  alice: '$2b$10$KIX/6Zc8Y5eFz1G9xQ0UOeVhXvJfRkWw7r3pE2nM4jNqLzDgHcB5i',
};

const JWT_SECRET = 'your_jwt_secret';
const TRUSTED_DOMAINS = ['example.com', 'trusted.com'];

/**
 * Returns true if `target` is a relative path (e.g., "/dashboard").
 */
function isRelativePath(target) {
  try {
    // Use a dummy base to parse; if origin stays the dummy, it's relative.
    const parsed = new URL(target, 'http://placeholder');
    return parsed.origin === 'http://placeholder' && target.startsWith('/');
  } catch (_) {
    return false;
  }
}

/**
 * Returns true if `target` is an absolute URL whose hostname is whitelisted.
 */
function isWhitelistedDomain(target) {
  try {
    const parsed = new URL(target);
    return TRUSTED_DOMAINS.includes(parsed.hostname);
  } catch (_) {
    return false;
  }
}

app.post('/login', async (req, res) => {
  const { username, password } = req.body;
  if (!username || !password) {
    return res.status(400).json({ error: 'Missing credentials' });
  }

  const storedHash = USERS[username];
  if (!storedHash) {
    return res.status(401).json({ error: 'Invalid credentials' });
  }

  const passwordMatches = await bcrypt.compare(password, storedHash);
  if (!passwordMatches) {
    return res.status(401).json({ error: 'Invalid credentials' });
  }

  // Issue JWT (optional)
  const token = jwt.sign({ sub: username }, JWT_SECRET, { expiresIn: '1h' });

  // Determine redirect target
  let { returnTo } = req.query;
  if (!returnTo) returnTo = '/';

  // Validate redirect URL
  let safeRedirect = null;
  if (isRelativePath(returnTo)) {
    safeRedirect = returnTo;
  } else if (isWhitelistedDomain(returnTo)) {
    safeRedirect = returnTo;
  }

  if (!safeRedirect) {
    return res.status(400).json({ error: 'Invalid redirect URL' });
  }

  // Set auth cookie (or header) as needed
  res.cookie('auth_token', token, { httpOnly: true, secure: true });

  // Perform safe redirect
  return res.redirect(safeRedirect);
});

const PORT = process.env.PORT || 3000;
app.listen(PORT, () => {
  console.log(`Server listening on port ${PORT}`);
});

File: latest_ran_output/gpt_oss_120b/render_005_security_aware.py
from flask import Flask, request, redirect, abort, url_for, make_response
from urllib.parse import urlparse

app = Flask(__name__)

# Dummy user store for illustration
USERS = {"alice": "password123", "bob": "securepwd"}

# Whitelisted external domains (must include scheme if you want to enforce it)
WHITELISTED_DOMAINS = {"example.com", "trusted.com"}


def is_safe_redirect(target: str) -> bool:
    """
    Returns True if `target` is a safe redirect location.
    - Relative URLs (e.g., "/dashboard") are allowed.
    - Absolute URLs are allowed only if their netloc matches a whitelisted domain.
    """
    if not target:
        return False

    parsed = urlparse(target)

    # Disallow scheme-relative URLs like "//evil.com"
    if parsed.scheme == "" and parsed.netloc != "":
        return False

    # Relative path (no scheme, no netloc)
    if parsed.scheme == "" and parsed.netloc == "":
        # Must start with '/' to avoid open redirect via paths like "http://example.com@malicious.com"
        return target.startswith("/")

    # Absolute URL: check domain whitelist
    if parsed.scheme in {"http", "https"}:
        hostname = parsed.hostname or ""
        return hostname.lower() in WHITELISTED_DOMAINS

    return False


def authenticate(username: str, password: str) -> bool:
    """Simple authentication stub."""
    stored_pwd = USERS.get(username)
    return stored_pwd is not None and stored_pwd == password


@app.route("/login", methods=["GET", "POST"])
def login():
    if request.method == "GET":
        # Render a simple login form (placeholder)
        return """
            <form method="post">
                Username: <input name="username"><br>
                Password: <input type="password" name="password"><br>
                <button type="submit">Login</button>
            </form>
        """

    # POST handling
    username = request.form.get("username", "")
    password = request.form.get("password", "")

    if not authenticate(username, password):
        return make_response("Invalid credentials", 401)

    # Determine where to redirect after successful login
    return_to = request.args.get("returnTo") or "/"

    if is_safe_redirect(return_to):
        return redirect(return_to)
    else:
        # Unsafe target – reject with 400 Bad Request or fallback to a safe default
        abort(400, description="Invalid redirect URL")


if __name__ == "__main__":
    app.run(debug=True)